scripts/bert/fp16_utils.py

```diff
@@ -0,0 +1,242 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""Trainer for mixed precision training."""
+import logging
+import numpy as np
+import mxnet as mx
+from mxnet import nd
+
+def grad_global_norm(parameters, max_norm):
+    """Calculate the 2-norm of gradients of parameters, and how much they should be scaled down
+    such that their 2-norm does not exceed `max_norm`.
+
+    If gradients exist for more than one context for a parameter, user needs to explicitly call
+    ``trainer.allreduce_grads`` so that the gradients are summed first before calculating
+    the 2-norm.
+
+    .. note::
+
+        This function is only for use when `update_on_kvstore` is set to False in trainer.
+
+    Example::
+
+        trainer = Trainer(net.collect_params(), update_on_kvstore=False, ...)
+        for x, y in mx.gluon.utils.split_and_load(X, [mx.gpu(0), mx.gpu(1)]):
+            with mx.autograd.record():
+                y = net(x)
+                loss = loss_fn(y, label)
+            loss.backward()
+        trainer.allreduce_grads()
+        norm, ratio = grad_global_norm(net.collect_params().values(), max_norm)
+        trainer.update(batch_size * ratio)
+        ...
+
+    Parameters
+    ----------
+    parameters : list of Parameters
+
+    Returns
+    -------
+    NDArray
+      Total norm.
+    NDArray
+      Ratio for rescaling gradients based on max_norm s.t. grad = grad / ratio.
+      If total norm is NaN, ratio will be NaN, too.
+    """
+    # collect gradient arrays
+    arrays = []
+    idx = 0
+    for p in parameters:
+        if p.grad_req != 'null':
+            p_grads = p.list_grad()
+            arrays.append(p_grads[idx % len(p_grads)])
+            idx += 1
+    assert len(arrays) > 0, 'No parameter found available for gradient norm.'
+
+    # compute gradient norms
+    def _norm(array):
+        # TODO(haibin) norm operator does not support fp16 safe reduction.
+        # Issue is tracked at: https://github.com/apache/incubator-mxnet/issues/14126
+        x = array.reshape((-1,)).astype('float32', copy=False)
+        return nd.dot(x, x)
+    norm_arrays = [_norm(arr) for arr in arrays]
+
+    # group norm arrays by ctx
+    def group_by_ctx(arr_list):
+        groups = {}
+        for arr in arr_list:
+            ctx = arr.context
+            if ctx in groups:
+                groups[ctx].append(arr)
+            else:
+                groups[ctx] = [arr]
+        return groups
+    norm_groups = group_by_ctx(norm_arrays)
+
+    # reduce
+    ctx, dtype = arrays[0].context, 'float32'
+    total_norm = mx.nd.zeros((1,), ctx=ctx, dtype=dtype)
+    for group in norm_groups:
+        total_norm += nd.add_n(*norm_groups[group]).as_in_context(ctx)
```

Comment:
nit: maybe use add_n again to avoid the sequential add?

Reply:
e.g. `total_norm = nd.add_n(*(nd.add_n(*ng).as_in_context(ctx) for ng in norm_groups.values()))`